### PR TITLE
Critical Constraint Modeling [WIP]

### PR DESCRIPTION
**Pull request by aaswenson**
What this PR contains:

1. Reworks the calculation to include a critical core radius constraint 
2. Improves material property handling 
3. Includes Bessel function approximation for radial flux profile (in addition to the axial cosine shape)
4. Some updated commenting, removal of unnecessary variables 
5. Updated unit tests

WIP:

1. Include cladding in the resistance network
2. Write unit tests for the nusselt correlations taken from EES.

### CODE REVIEW

File: optimization/ht_functions.py [hunk 1]
```diff
@@ -1,12 +1,128 @@
...
+    Gz = Re* Pr /LD     
+    x = LD / Re
+    fR = 3.44 / math.sqrt(x) +\
+        (1.25/(4*x) + 16 - 3.44/math.sqrt(x)) /\
+        (1 + 0.00021 * x**(-2))
+
+    f = 4 * fR / Re
+    Gm = Gz**(1/3)
+    Nusselt_T = 3.66 + ((0.049+0.02/Pr)*Gz**1.12)/(1+0.065*Gz**0.7)
+    Nusselt_H = 4.36 + ((0.1156 +0.08569 /Pr**0.4)*Gz)/(1+0.1158*Gz**0.6)
```
Comment (bam241) [inline optimization/ht_functions.py:82]: maybe you could had some refs for those ? So those magic number could make some sense

File: optimization/ht_functions.py [hunk 2]
```diff
@@ -1,12 +1,128 @@
...
+        Re (float): flow Reynold's number [-]
+        Pr (float): flow Prandtl number [-]
+        LD (float): length over diameter [-]
+        relrough (float) : relative roughness [-]
+    
+    Returns:
+    --------
+        Nusselt_L (float): nusselt number
+        f (float): friction factor
+    """
```
Comment (bam241) [inline optimization/ht_functions.py:21]: Maybe you should add some test there to ensure this is not used on a laminar case ?
(as it seems to only depends on `Re`)

File: optimization/ht_functions.py [hunk 3]
```diff
@@ -1,12 +1,128 @@
...
+        else:
+            Nusselt_L = Nusselt_L_lp+(Pr-0.1)*(Nusselt_L-Nusselt_L_lp)/0.4
+
+    #account for developing flow
+    f=f_fd*(1+(1/LD)**0.7) 
+    Nusselt_L*=(1+(1/LD)**0.7)
+    
+    return Nusselt_L, f
+
+def pipeflow_laminar(Re, Pr, LD, relrough):
```
Comment (bam241) [inline optimization/ht_functions.py:55]: same here to ensure the laminar regime is actually true ?
(as it seems to only depends on `Re`)

File: optimization/ht_functions.py [hunk 4]
```diff
@@ -1,12 +1,128 @@
...
+    
+    Returns:
+    --------
+        Nusselt_T (float): nusselt number at constant temperature
+        Nusselt_H (float): nusselt number at constant heat flux
+        f (float): friction factor
+    """
+    if Re > 3000:
+        Nusselt_T, f = pipeflow_turbulent(Re, Pr, LD, relrough)
+        Nusselt_H = Nusselt_T
```
Comment (bam241) [inline optimization/ht_functions.py:106]: I don't like that, I would prefere if both pipe flow_xxx return all of `Nusselt_T`, `Nusselt_H`, `f`... (with `Nusselt_H = Nusselt_T` for the turbulent case.

File: optimization/ht_functions.py [hunk 5]
```diff
@@ -1,12 +1,128 @@
...
+        (1 + 0.00021 * x**(-2))
+
+    f = 4 * fR / Re
+    Gm = Gz**(1/3)
+    Nusselt_T = 3.66 + ((0.049+0.02/Pr)*Gz**1.12)/(1+0.065*Gz**0.7)
+    Nusselt_H = 4.36 + ((0.1156 +0.08569 /Pr**0.4)*Gz)/(1+0.1158*Gz**0.6) 
+    
+    return Nusselt_T, Nusselt_H, f
+    
+def pipeflow_nd(Re, Pr, LD, relrough):
```
Comment (bam241) [inline optimization/ht_functions.py:86]: what does the nd mean ? why not simply `pipeflow` ?

File: optimization/ht_functions.py [hunk 6]
```diff
@@ -58,77 +173,55 @@ def find_n_channels(flow):
...
 
         Initialized Attributes:
         --------------------
-            r_channel: radius of coolant channel [m]
-            c: cladding thickness [m]
+            Q_therm (float): core thermal power [W]
+            fuel (str): reactor fuel
+            coolant (str): reactor coolant
+            AR (float): core aspect ratio [-]
+            c (float): clad thickness
```
Comment (bam241) [inline optimization/ht_functions.py:203]: units ?

File: optimization/ht_functions.py [hunk 7]
```diff
@@ -58,77 +173,55 @@ def find_n_channels(flow):
...
+            r_channel (float): coolant channel radius
```
Comment (bam241) [inline optimization/ht_functions.py:204]: units?

File: optimization/ht_functions.py [hunk 8]
```diff
@@ -182,44 +296,31 @@ def get_q_per_channel(self):
...
-            R_conv: resistance term (convection to the fluid) [W/K]
-            R_tot: total resistance to HT [W/K]
-            q_per_channel: total generation in fuel channel [W]
-            q_bar: axially-averaged volumetric generation in fuel [W]
-            N_channels: required channels for desired Q [-]
+            radius_cond (float): average distance of conduction in fuel [m]
+            R_fuel: (float): resistance term (conduction in the fuel) [W/K]
+            R_conv: (float): resistance term (convection to the coolant) [W/K]
+            q_trip_max (float): maximum achievable thermal power [W]
+            gen_Q (float): thermal power scaled for flux shape
```
Comment (bam241) [inline optimization/ht_functions.py:308]: not sure I understand that one, does it have units ?

File: optimization/ht_functions.py [hunk 9]
```diff
@@ -303,86 +393,30 @@ def calc_reactor_mass(self):
...
-        # sweep through parameter space, calculate min mass
-        for i in range(self.N):
-            for j in range(self.N):
-                flowdata = Flow(R_mesh[i, j], PD_mesh[i, j], c, z, props)
-                oned_flow_modeling(flowdata)
-                self.save_iteration(flowdata, i, j)
-
+        fuel_mass = self.vol_fuel * self.fuelprops['rho_fuel']
+        cool_mass = self.vol_cool * self.fps.rho
+        refl_mass = ((self.core_r * 1.05)**2 - self.core_r**2) *\
```
Comment (bam241) [inline optimization/ht_functions.py:409]: `1.05`: magic number ?

File: optimization/ht_functions.py [hunk 10]
```diff
@@ -303,86 +393,30 @@ def calc_reactor_mass(self):
...
-        self.minPD = self.data[self.min_idx]['pd']
-        # report the optimal configuration and it's corresponding fuel mass
-        outstring = "1D Thermal Hydraulics Optimization Results:\n"
-        outstring += "Min reactor mass (m = " + str(round(self.min_mass, 3))\
-            + "[kg]) " + " occurs at r = " + str(self.minD) + "[m] & PD = "\
-            + str(self.minPD) + "[-]."
-        print(outstring)
+                   'UW'  : {'CO2' : (0.15385, -0.8309),
+                            'H2O' : (0.16270, -0.6487)
+                           }
```
Comment (bam241) [inline optimization/ht_functions.py:428]: could you tell more about those coefficients ?

File: optimization/reactor_mass.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+        coolant (str): coolant type
+        power (float): thermal power [W]
+        m_dot (float): mass flow rate [kg/s]
+        T ((float, float)): inlet, outlet temp [K]
+        P ((float, float)): inlet, outlet pressure [Pa]
+        clad (optional, str): cladding type, default is Inconel-718
+        refl (optional, str): reflector type, default is Carbon
+        cool_r (optional, float): coolant channel radius [m]
+        clad_t (optional, float): clad thickness [m]
+        AR (optional, float): core aspect ratio [-]
```
Comment (bam241) [inline optimization/reactor_mass.py:25]: is there a default values?

File: optimization/test_ht_functions.py [hunk 12]
```diff
@@ -1,81 +1,109 @@
...
     test.characterize_flow()
     # compare
-    assert exp_v == test.v
-    assert exp_f == test.f
-    assert exp_h == test.h_bar
-
+    assert abs(exp_v - test.v) < 1e-7
+    assert abs(exp_f - test.f) < 1e-7
+    assert abs(exp_Nu - test.Nu) < 1e-7
+    assert abs(exp_h - test.h_bar) < 1e-5
```
Comment (bam241) [inline optimization/test_ht_functions.py:76]: why not `1e-7` ?

Comment (aaswenson) [reply optimization/test_ht_functions.py:76]: The absolute error grows in the calculation for h_bar

File: optimization/physical_properties.py [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+import numpy as np
+from scipy import interpolate
+
+kJ_to_J = 1e3
+
+fuel_props = {'UW' : {
+                'T_center'  : 1847.5, # centerline fuel temp (max) [K]
+                'rho_W'     : 19300,  # clad density [kg/m^3]
+                'rho_UN'    : 11300,  # fuel density [kg/m^3]
+                'fuel_frac' : 0.6,    # volume fraction of fuel in CERMET
```
Comment (bam241) [inline optimization/physical_properties.py:13]: ref ?

**Pull request merged**